x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/shards/FrozenShardsDeciderService.java

```diff
@@ -0,0 +1,141 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.autoscaling.shards;
+
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingCapacity;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderContext;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderResult;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderService;
+import org.elasticsearch.xpack.cluster.routing.allocation.DataTierAllocationDecider;
+import org.elasticsearch.xpack.core.DataTier;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.StreamSupport;
+
+/**
+ * This decider enforces that on a 64GB memory node (31GB heap) we can max have 2000 shards. We arrive at 2000 because our current limit is
+ * 1000 but frozen tier uses the "frozen engine", which is much more efficient. We scale the total tier memory accordingly.
+ *
+ * The decider relies on frozen tier being used exclusively for frozen shards.
+ */
+public class FrozenShardsDeciderService implements AutoscalingDeciderService {
+    public static final String NAME = "frozen_shards";
+    private static final ByteSizeValue MAX_MEMORY = ByteSizeValue.ofGb(64);
+    static final ByteSizeValue DEFAULT_MEMORY_PER_SHARD = ByteSizeValue.ofBytes(MAX_MEMORY.getBytes() / 2000);
+    public static final Setting<ByteSizeValue> MEMORY_PER_SHARD = Setting.byteSizeSetting(
+        "memory_per_shard",
+        (dummy) -> DEFAULT_MEMORY_PER_SHARD.getStringRep(),
+        ByteSizeValue.ZERO,
+        ByteSizeValue.ofBytes(Long.MAX_VALUE)
+    );
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public AutoscalingDeciderResult scale(Settings configuration, AutoscalingDeciderContext context) {
+        // we assume that nodes do not grow beyond 64GB here.
+        long shards = countFrozenShards(context.state().metadata());
+        long memory = shards * MEMORY_PER_SHARD.get(configuration).getBytes();
+        return new AutoscalingDeciderResult(AutoscalingCapacity.builder().total(null, memory).build(), new FrozenShardsReason(shards));
+    }
+
+    static long countFrozenShards(Metadata metadata) {
+        return StreamSupport.stream(metadata.spliterator(), false)
+            .filter(imd -> isFrozenIndex(imd.getSettings()))
+            .mapToLong(IndexMetadata::getTotalNumberOfShards)
```

Comment:
I’m curious why `mapToLong` instead of `mapToInt` here, and `IntStream#sum` returning an `int` then?

Reply:
future proofing :slightly_smiling_face:, avoiding to think about (int*long) and thinking we only support 64bit so it does not matter. I changed it to an int in 6c5ecdebff415b69f708cfec704351aeb3a73d62 since this seems to be the "convention" for such a count.